### PR TITLE
[#391] Add support for json structured output for vLLM

### PR DESCRIPTION
**Pull request by aidando73**
# What does this PR do?

Addresses issue (#391)

- Adds json structured output for vLLM
- Enables structured output tests for vLLM

> Give me a recipe for Spaghetti Bolognaise:

```json
{
  "recipe_name": "Spaghetti Bolognaise",
  "preamble": "Ah, spaghetti bolognaise - the quintessential Italian dish that fills my kitchen with the aromas of childhood nostalgia. As a child, I would watch my nonna cook up a big pot of spaghetti bolognaise every Sunday, filling our small Italian household with the savory scent of simmering meat and tomatoes. The way the sauce would thicken and the spaghetti would al dente - it was love at first bite. And now, as a chef, I want to share that same love with you, so you can recreate these warm, comforting memories at home.",
  "ingredients": [
    "500g minced beef",
    "1 medium onion, finely chopped",
    "2 cloves garlic, minced",
    "1 carrot, finely chopped",
    " celery, finely chopped",
    "1 (28 oz) can whole peeled tomatoes",
    "1 tbsp tomato paste",
    "1 tsp dried basil",
    "1 tsp dried oregano",
    "1 tsp salt",
    "1/2 tsp black pepper",
    "1/2 tsp sugar",
    "1 lb spaghetti",
    "Grated Parmesan cheese, for serving",
    "Extra virgin olive oil, for serving"
  ],
  "steps": [
    "Heat a large pot over medium heat and add a generous drizzle of extra virgin olive oil.",
    "Add the chopped onion, garlic, carrot, and celery and cook until the vegetables are soft and translucent, about 5-7 minutes.",
    "Add the minced beef and cook until browned, breaking it up with a spoon as it cooks.",
    "Add the tomato paste and cook for 1-2 minutes, stirring constantly.",
    "Add the canned tomatoes, dried basil, dried oregano, salt, black pepper, and sugar. Stir well to combine.",
    "Bring the sauce to a simmer and let it cook for 20-30 minutes, stirring occasionally, until the sauce has thickened and the flavors have melded together.",
    "While the sauce cooks, bring a large pot of salted water to a boil and cook the spaghetti according to the package instructions until al dente. Reserve 1 cup of pasta water before draining the spaghetti.",
    "Add the reserved pasta water to the sauce and stir to combine.",
    "Combine the cooked spaghetti and sauce, tossing to coat the pasta evenly.",
    "Serve hot, topped with grated Parmesan cheese and a drizzle of extra virgin olive oil.",
    "Enjoy!"
  ]
}
```

Generated with Llama-3.2-3B-Instruct model - pretty good for a 3B parameter model 👍

## Test Plan

`pytest -v -s llama_stack/providers/tests/inference/test_text_inference.py -k llama_3b-vllm_remote`

With the following setup:

```bash
# Environment
export INFERENCE_MODEL=meta-llama/Llama-3.2-3B-Instruct
export INFERENCE_PORT=8000
export VLLM_URL=http://localhost:8000/v1

# vLLM server
sudo docker run --gpus all \
    -v $STORAGE_DIR/.cache/huggingface:/root/.cache/huggingface \
    --env "HUGGING_FACE_HUB_TOKEN=$(cat ~/.cache/huggingface/token)" \
    -p 8000:$INFERENCE_PORT \
    --ipc=host \
    --net=host \
    vllm/vllm-openai:v0.6.3.post1 \
    --model $INFERENCE_MODEL

# llama-stack server
llama stack build --template remote-vllm --image-type conda && llama stack run distributions/remote-vllm/run.yaml \
  --port 5001 \
  --env INFERENCE_MODEL=meta-llama/Llama-3.2-3B-Instruct
```

Results:

```
llama_stack/providers/tests/inference/test_text_inference.py::TestInference::test_model_list[llama_3b-vllm_remote] PASSED
llama_stack/providers/tests/inference/test_text_inference.py::TestInference::test_completion[llama_3b-vllm_remote] SKIPPED
llama_stack/providers/tests/inference/test_text_inference.py::TestInference::test_completions_structured_output[llama_3b-vllm_remote] SKIPPED
llama_stack/providers/tests/inference/test_text_inference.py::TestInference::test_chat_completion_non_streaming[llama_3b-vllm_remote] PASSED
llama_stack/providers/tests/inference/test_text_inference.py::TestInference::test_structured_output[llama_3b-vllm_remote] PASSED
llama_stack/providers/tests/inference/test_text_inference.py::TestInference::test_chat_completion_streaming[llama_3b-vllm_remote] PASSED
llama_stack/providers/tests/inference/test_text_inference.py::TestInference::test_chat_completion_with_tool_calling[llama_3b-vllm_remote] PASSED
llama_stack/providers/tests/inference/test_text_inference.py::TestInference::test_chat_completion_with_tool_calling_streaming[llama_3b-vllm_remote] PASSED

================================ 6 passed, 2 skipped, 120 deselected, 2 warnings in 13.26s ================================
```

## Sources

- https://github.com/vllm-project/vllm/discussions/8300
- By default, vLLM uses https://github.com/dottxt-ai/outlines for structured outputs [[1](https://github.com/vllm-project/vllm/blob/32e7db25365415841ebc7c4215851743fbb1bad1/vllm/engine/arg_utils.py#L279-L280)]

## Before submitting

[N/A] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case)

- [x] Ran pre-commit to handle lint / formatting issues.
- [x] Read the [contributor guideline](https://github.com/meta-llama/llama-stack/blob/main/CONTRIBUTING.md),
      Pull Request section?

[N/A?] Updated relevant documentation. Couldn't find any relevant documentation. Lmk if I've missed anything.

- [x] Wrote necessary unit or integration tests.

### CODE REVIEW

File: llama_stack/providers/remote/inference/vllm/vllm.py [hunk 1]
```diff
@@ -180,6 +181,16 @@ async def _get_params(
                 self.formatter,
             )
 
+        if fmt := request.response_format:
+            if fmt.type == ResponseFormatType.json_schema.value:
+                input_dict["extra_body"] = {
+                    "guided_json": request.response_format.json_schema
+                }
```
Comment (aidando73) [inline llama_stack/providers/remote/inference/vllm/vllm.py:188]: Note that using

```python
input_dict["response_format"] = {
  "type": "json_schema",
  "json_schema": {
    "name": "example_name",
    "schema": request.response_format.json_schema
  }
}
```

as per the usual OpenAI api, doesn't work.

**Comment by aidando73** [comment]
### Script to play around with

```python
import os
from llama_stack_client import LlamaStackClient
from pydantic import BaseModel

client = LlamaStackClient(base_url=f"http://localhost:{os.environ['LLAMA_STACK_PORT']}")

class CompletionMessage(BaseModel):
    content: str
    additional_info: str

response = client.inference.chat_completion(
    model_id=os.environ["INFERENCE_MODEL"],
    messages=[
        {"role": "system", "content": "You are a helpful assistant."},
        {"role": "user", "content": "Write a haiku about coding"},
    ],
    response_format={
        "type": "json_schema",
        "json_schema": CompletionMessage.model_json_schema(),
    },
)
print(response.completion_message.content)
```

File: llama_stack/providers/remote/inference/vllm/vllm.py [hunk 2]
```diff
@@ -180,6 +181,16 @@ async def _get_params(
...
+            elif fmt.type == ResponseFormatType.grammar.value:
+                raise NotImplementedError("Grammar response format not supported yet")
+            else:
+                raise ValueError(f"Unknown response format {fmt.type}")
+
         return {
             "model": request.model,
             **input_dict,
```
Comment (aidando73) [inline llama_stack/providers/remote/inference/vllm/vllm.py:196]: Prompt adaptation also working, our prompts are being adapted to:

```

```

Becomes:

```
<|begin_of_text|><|start_header_id|>system<|end_header_id|>

You are a helpful assistant.<|eot_id|><|start_header_id|>user<|end_header_id|>

Write a haiku about coding<|eot_id|><|start_header_id|>user<|end_header_id|>

Please respond in JSON format with the schema: {properties: {content: {title: Content, type: string}, additional_info: {title: Additional Info, type: string}}, required: [content, additional_info], title: CompletionMessage, type: object}<|eot_id|><|start_header_id|>assistant<|end_header_id|>
```

File: llama_stack/providers/remote/inference/vllm/vllm.py (hunk 1 above)
Comment (terrytangyuan) [reply llama_stack/providers/remote/inference/vllm/vllm.py:188]: Can you document it in an vLLM issue and link it back here?

**Review by terrytangyuan — COMMENTED** [review]
Thank you! 

Can you submit the existing issue in vLLM repo and link it back here as code comments?

**Comment by aidando73** [comment]
@terrytangyuan Sure:

https://github.com/vllm-project/vllm/issues/7220 (existing issue)

File: llama_stack/providers/tests/inference/test_text_inference.py [hunk 3]
```diff
@@ -140,7 +140,9 @@ async def test_completions_structured_output(
             "remote::tgi",
             "remote::together",
             "remote::fireworks",
+            "remote::vllm",
             "remote::cerebras",
+
```
Comment (aidando73) [inline llama_stack/providers/tests/inference/test_text_inference.py:145]:
```suggestion
```

**Review by terrytangyuan — APPROVED** [review]
LGTM. Thank you!

**Comment by ashwinb** [comment]
🎉 🎉 🎉 

Thank you!!

**Pull request merged**